core/src/query.rs

```diff
@@ -1,77 +1,551 @@
+                datasource_schema
+                    .clone()
+                    .properties
+                    // Since we are flat-mapping, we cannot conveniently return
+                    // a Result from this closure.  Therefore, we do not assert
+                    // that the datasource_schema is valid at this time, saving
+                    // that for the follow-up map. We proceed by assuming there
+                    // are properties,  or by using an empty HashMap when there
+                    // are not.  This is equivalent to handling "no properties"
+                    // without the need to return an error at this stage.
+                    .unwrap_or_else(|| HashMap::default())
+                    .into_iter()
+                    .sorted_by(|a, b| Ord::cmp(&a.0, &b.0))
+                    .map(move |(field_name, field_schema)| {
+                        (
+                            datasource_name.clone(),
+                            datasource_schema.clone(),
+                            field_name.clone(),
+                            field_schema.clone(),
+                        )
+                    })
+            })
+            // 3. Map each field into a MongoColMetadata.
+            .map(
+                |(datasource_name, datasource_schema, field_name, field_schema)| {
+                    datasource_schema.assert_datasource_schema()?;
+
+                    let field_nullability =
+                        datasource_schema.get_field_nullability(field_name.clone())?;
+
+                    Self::create_column_metadata(
+                        current_db,
+                        datasource_name,
+                        field_name,
+                        field_schema,
+                        field_nullability,
+                    )
+                },
+            )
+            // 4. Collect as a Vec.
+            .collect::<Result<Vec<MongoColMetadata>>>()
+    }
+
+    fn create_column_metadata(
+        current_db: &String,
+        datasource_name: String,
+        field_name: String,
+        field_schema: SimplifiedJsonSchema,
+        is_nullable: ColumnNullability,
+    ) -> Result<MongoColMetadata> {
+        let bson_type_info: BsonTypeInfo = (field_name.clone(), field_schema).try_into()?;
+
+        Ok(MongoColMetadata {
+            // For base_col_name and base_table_name, we do not have this
+            // information in sqlGetResultSchema, so this will always be
+            // empty string.
+            base_col_name: "".to_string(),
+            base_table_name: "".to_string(),
+            // For catalog_name, we do not have this information in
+            // sqlGetResultSchema, so this will always be current_db. This
+            // is not correct for correct for fields from tables in other
+            // databases as part of cross-db lookups, but this is the best
+            // we can do for now.
+            catalog_name: current_db.clone(),
```

Comment:
Given the comment above, do we think this will be a problem?

Reply:
Is there any reason we can't just make it the empty string?

Reply:
We definitely can for now. That's the safer choice. I'll make that change.